File: solutions/cpp/log-levels/1/log_levels.cpp
#include <string>

namespace log_line {
    std::string message(std::string line) {
        size_t pos = line.find(" ");
        if (pos != std::string::npos) {
            return line.substr(pos + 1);
        }
        return "";
    }
    
    std::string log_level(std::string line) {
        size_t pos_begin = line.find("[");
        if (pos_begin != std::string::npos) {
            size_t pos_end = line.find("]: ");
            if (pos_end != std::string::npos) {
                return line.substr(pos_begin+1, pos_end-pos_begin-1);
            }
        }
        return "";
    }
    
    std::string reformat(std::string line) {
        return message(line) + " (" + log_level(line) + ")";
    }
}  // namespace log_line

File: solutions/csharp/log-levels/1/LogLevels.cs
static class LogLine
{
    public static string Message(string logLine)
    {
        int idx = logLine.IndexOf("]: ");
        if (idx >= 0) {
            return logLine.Substring(idx+3).Trim();
        }
        return logLine.Trim();
    }

    public static string LogLevel(string logLine)
    {
        int endIdx = logLine.IndexOf("]: ");
        if (endIdx >= 0) { 
            int startIdx = logLine.IndexOf("[");
            if (startIdx >= 0 && startIdx < endIdx) {
                startIdx += 1;
                int len = endIdx - startIdx;
                return logLine.Substring(startIdx, len).ToLower();
            }
        }
        return logLine;
    }

    public static string Reformat(string logLine) => $"{Message(logLine)} ({LogLevel(logLine)})";
}
